Orchestrator/Prognetics.Saga.Orchestrator/SagaEngine.cs

```diff
@@ -1,33 +1,125 @@
+using Microsoft.Extensions.Logging;
 using Prognetics.Saga.Core.Abstract;
+using Prognetics.Saga.Core.Model;
 using Prognetics.Saga.Orchestrator.Contract;
 using Prognetics.Saga.Orchestrator.Contract.DTO;
+using System.Transactions;
 
 namespace Prognetics.Saga.Orchestrator;
 
 public class SagaEngine : ISagaEngine
 {
+    private readonly ISagaLog _sagaLog;
+    private readonly ICompensationStore _compensationStore;
     private readonly ITransactionLedgerAccessor _transactionLedgerAccessor;
+    private readonly IdGenerator _idGenerator;
+    private readonly ILogger<ISagaEngine> _logger;
 
-    public SagaEngine(ITransactionLedgerAccessor transactionLedgerAccessor)
+    public SagaEngine(
+        ISagaLog sagaLog,
+        ICompensationStore compensationStore,
+        ITransactionLedgerAccessor transactionLedgerAccessor,
+        IdGenerator idGenerator,
+        ILogger<ISagaEngine> logger)
     {
+        _sagaLog = sagaLog;
+        _compensationStore = compensationStore;
         _transactionLedgerAccessor = transactionLedgerAccessor;
+        _idGenerator = idGenerator;
+        _logger = logger;
     }
 
-    public Task<EngineOutput?> Process(EngineInput input)
+    public async Task<EngineOutput?> ProcessOrDefault(EngineInput input)
     {
         var transactionLedger = _transactionLedgerAccessor.TransactionsLedger;
-        var transactionModel = transactionLedger.GetTransactionByCompletionEventName(input.EventName);
-        var stepRecord = transactionModel?.GetStepByCompletionEventName(input.EventName);
+        var transactionStep = transactionLedger.GetTransactionStepByCompletionEventName(input.EventName);
 
-        if (stepRecord is null || transactionModel is null)
+        if (transactionStep is null)
         {
-            return Task.FromResult((EngineOutput?)null);
+            _logger.LogError("Unknown event name: {EventName}", input.EventName);
+            return null;
         }
 
-        return Task.FromResult((EngineOutput?)new EngineOutput(
-            stepRecord.Step.CompletionEventName,
-            new OutputMessage(
-                input.Message.TransactionId ?? Guid.NewGuid().ToString(),
-                input.Message.Payload)));
+        var transactionId = input.Message.TransactionId;
+
+        if (transactionStep.IsFirst)
+        {
+            transactionId = await InitializeTransaction(transactionStep);
+        }
+        else if (transactionId is null)
+        {
+            _logger.LogError("Transaction id is missing");
+            return null;
+        }
+        else if (await TryProcessExistingTransaction(transactionId, transactionStep) == false)
```

Comment:
Order of those if's is hardly readable. It's hard to tell why next steps occur in this specific way. Can we maybe split this into more methods or use some pattern?